backward.hpp

```diff
@@ -811,13 +811,63 @@ class TraceResolverImplBase {
 
 #ifdef BACKWARD_SYSTEM_LINUX
 
+static std::string get_argv0() {
+  std::string argv0;
+  std::getline(std::ifstream("/proc/self/cmdline"), argv0, '\0');
+  return argv0;
+}
+
+static std::string read_symlink(std::string const &symlink_path) {
+  std::string path;
+  path.resize(100);
+
+  while (true) {
+    ssize_t len =
+        ::readlink(symlink_path.c_str(), &*path.begin(), path.size());
+    if (len < 0) {
+      return "";
+    }
+    if (static_cast<size_t>(len) == path.size()) {
+      path.resize(path.size() * 2);
+    } else {
+      path.resize(static_cast<std::string::size_type>(len));
+      break;
+    }
+  }
+
+  return path;
+}
+
+class TraceResolverLinuxBase
+    : public TraceResolverImplBase {
+public:
+  void fixup_dladdr_filenames(Dl_info & symbol_info, std::string & display_filename) {
+    // returns filename to open in symbol_info.dli_fname and filename to display in display_filename
+    if(symbol_info.dli_fname == argv0_) {
+      // dladdr returns argv[0] in dli_fname for symbols contained in
+      // the main executable, which is not a valid path if the
+      // executable was found by a search of the PATH environment
+      // variable; In that case, we actually open /proc/self/exe, which
+      // is always the actual executable (even if it was deleted/replaced!)
+      // but display the path that /proc/self/exe links to.
+      symbol_info.dli_fname = "/proc/self/exe";
+      display_filename = exec_path_;
+    } else {
+      display_filename = symbol_info.dli_fname;
+    }
+  }
+private:
+  std::string argv0_ = get_argv0();
```

Comment:
this syntax doesn't work on C++98. You can use a list initializer in the constructor.